bug: macos-latest consumer-install smoke crashes with libc++ mutex error

## Summary

Consumer-install smoke harness crashes on every `macos-latest` CI run with a native libc++ mutex error before any moflo code runs. **This is a pre-existing CI breakage on `main`** (failed on 12:24Z `main` run prior to PR #612 even being opened), and has been silently red on `macos-latest` for at least the most recent push to main. **A first attempt to fix this landed on branch `fix/613-onnxruntime-node-macos-mutex` (PR #615) but does NOT solve the macOS crash — see "Prior attempts" below before re-trying.**

## Reproduction

CI: `Consumer install smoke / macos-latest` (any recent run).

Failing runs:
- main: https://github.com/eric-cielo/moflo/actions/runs/24930845358
- PR #615 (failed fix attempt): https://github.com/eric-cielo/moflo/actions/runs/24932238217

Runner image: `macos-15-arm64/20260421.0007` (macOS 15.7.4, Node 20.20.2).

## Symptom

Every `flo` subprocess that loads `fastembed`/`onnxruntime-node` dies at process exit with:

```
libc++abi: terminating due to uncaught exception of type std::__1::system_error:
  mutex lock failed: Invalid argument
```

12 of 38 smoke checks fail per run. Affected: `doctor`, `memory-init/store/retrieve/search/list/delete`, `spell-list`, `hooks-list/pre-task/post-edit`, plus `check` (cascades from memory-init).

PASSES (don't load ORT in-process): `flo --version`, `flo mcp tools`, `flo-search`, the `moflodb-bridge` probe, all `probe:load:*`. `ubuntu-latest` and `windows-latest` pass clean.

## Root cause (CONFIRMED — do not re-investigate)

Upstream [microsoft/onnxruntime#24579](https://github.com/microsoft/onnxruntime/issues/24579) — "mutex issue at process exit on MacOS since v1.21.*".

`onnxruntime-node` 1.21.0 introduced static `std::mutex` instances inside the logger that get destroyed during `__cxa_finalize_ranges` when `OrtEnv` is torn down. On macOS, libc++'s mutex destructor is non-trivial (unlike libstdc++), so once the static mutex storage has been finalized, any subsequent destructor that touches it aborts with `EINVAL`. The bug is platform-agnostic in principle but only manifests on Apple targets.

**Fix lives in `onnxruntime-node@1.24.1+`** (PR [microsoft/onnxruntime#26445](https://github.com/microsoft/onnxruntime/pull/26445), merged 2025-10-31, first released in v1.24.1 on 2026-02-06). The fix moves the offending statics into function-local statics so the compiler emits the right destruction order. Verified the fix commit is in v1.24.1 via `git log v1.23.2..v1.24.1`.

The blocker: **fastembed@2.1.0 declares `onnxruntime-node@1.21.0` as an EXACT pin** (not a range). We need that pin relaxed.

## CRITICAL INSIGHT — npm overrides do NOT propagate to consumers

**This is the trap that broke PR #615. Future attempts must internalize this before touching code.**

- `overrides` in moflo's `package.json` apply **only when moflo itself is the active project root** (i.e., during moflo's own dev install / CI build).
- When a consumer runs `npm install moflo`, npm reads the consumer's package.json (which is just `{ devDependencies: { moflo: "file:..." } }` in the smoke harness, or `{ dependencies: { moflo: "..." } }` for a real consumer). The consumer's package.json has no overrides for ORT.
- npm then resolves moflo's transitive deps using the SHIPPED package metadata. fastembed's exact pin `onnxruntime-node@1.21.0` wins, and the broken version gets installed nested under `node_modules/fastembed/node_modules/onnxruntime-node/`.
- moflo's `overrides` block never even gets read by the consumer's npm.

**Verification:** the smoke harness's consumer scratch dir was directly inspected after running PR #615's install:
```
node_modules/onnxruntime-node/package.json → "version": "1.21.0"  (the broken one)
node_modules/onnxruntime-node/bin/napi-v3/   (1.21.0 layout)
```

Same dev install with `overrides`:
```
node_modules/onnxruntime-node/package.json → "version": "1.24.3"  (fixed)
node_modules/onnxruntime-node/bin/napi-v6/   (1.24.x layout)
```

The override "fixed" moflo's dev install only. **The macos-latest smoke runs the consumer install path** — that's why the override-only PR is still red.

## Prior attempts and why they failed

### Attempt 1 (LANDED on branch, in PR #615, does NOT fix macOS): npm `overrides` only

Commit `56d211ee5` on `fix/613-onnxruntime-node-macos-mutex`:
- Added `"onnxruntime-node": ">=1.24.3"` to root `package.json` `overrides` block.
- Updated `scripts/prune-native-binaries.mjs` and `harness/consumer-smoke/lib/checks.mjs` to discover `bin/napi-v<N>/` dynamically (since 1.24+ ships `napi-v6` instead of `napi-v3`).
- Added unit tests for the napi-v6 / mixed-tier cases.

**Status:** the override doesn't reach consumers (see CRITICAL INSIGHT above). macos-latest stays red.

**Salvage value:** the `napi-v<N>` discovery is real and worth keeping — the moment fastembed eventually relaxes its pin, our prune would silently no-op without this fix because ORT 1.24+ uses napi-v6 and the old pruner had `napi-v3` hardcoded. Whoever takes #613 next should keep that part of PR #615 and rescope the PR title/description.

### Attempt 2 (in-progress, REVERTED — do not re-try without solving the problems below): direct dep + postinstall surgery

Tried adding `"onnxruntime-node": ">=1.24.3"` to moflo's `dependencies` (so it gets hoisted in consumer installs as a top-level dep), then having `scripts/prune-native-binaries.mjs` postinstall delete fastembed's nested `node_modules/fastembed/node_modules/onnxruntime-node/` so fastembed's `import('onnxruntime-node')` falls back to the hoisted (fixed) copy.

This approach has THREE concrete problems that must be solved before retrying it:

1. **Install-size budget blown.** With both copies installed, consumer scratch dir hits 271.5 MB total (vs. 120 MB hard-fail threshold). Even after our postinstall pruner runs, both ORT copies exist for the duration of `npm install`, and the hoisted copy isn't pruned (because `findOrtPackages` already treats it as "shared with another declarer = moflo" and skips pruning by design — see ownership rules in `scripts/prune-native-binaries.mjs`).

2. **Windows: rmSync EPERM during postinstall.** `removeDir(nested, { maxRetries: 5, retryDelay: 500 })` still hits EPERM on Windows because file locks (likely Windows Defender or npm itself holding handles) prevent recursive directory removal. Falling back to deleting just `package.json` works for correctness (Node's resolution skips the dir), but the binaries still take ~80 MB and the install-size check fails.

3. **fastembed's nested `onnxruntime-node@1.21.0` has its OWN broken install.js on Windows.** Reproduced locally: `npm install` of the consumer scratch dir crashes with `npm error Error: Cannot find module ''` — an empty-string `require()` somewhere inside `onnxruntime-node@1.21.0`'s own `script/install.js`. This means the consumer install can't even COMPLETE on Windows, let alone reach our postinstall. Possibly fastembed-version-specific or platform-specific. Worth confirming whether this also affects ubuntu/macos before treating as a blocker (Windows is the platform that's currently green on smoke — it shouldn't be the one we just broke).

## Process-level lesson (what made this hard, must internalize)

Two analysis dimensions were under-considered before writing code:
- **Cross-platform**: Linux, macOS, Windows behave differently for native deps (libc++ vs libstdc++ vs MSVC; file locks; install.js postinstalls). A fix that works on Linux can crash on macOS or fail to install on Windows.
- **Consumer install**: a fix in moflo's package.json may only affect moflo's own dev tree. The smoke harness installs from a tarball into a scratch dir — that's the path real users see. Anything that depends on `overrides`, dev-only tooling, or moflo's project-root context will not propagate.

**Both dimensions must be answered explicitly in the ticket BEFORE writing any code** for any future attempt at this. The gate to enforce this on local dev was discussed but is OUT OF SCOPE for this issue and tracked separately.

## Paths forward (not yet tried, ranked by attractiveness)

### A. Upstream PR to fastembed to relax the pin (recommended first move)

Open a PR against [Anush008/fastembed-js](https://github.com/Anush008/fastembed-js) changing the `onnxruntime-node` declaration from exact `1.21.0` to a range like `^1.21.0 || ^1.24.1`. Their JS API surface is stable across this range; we already verified locally that fastembed@2.1.0 works against onnxruntime-node@1.24.3.

- **Pros:** correct fix; stops being moflo's problem permanently; benefits every fastembed consumer hitting the macOS crash.
- **Cons:** blocked on review/merge timing; macos-latest stays red until they ship.
- **Mitigation while blocked:** keep the napi-v6 prune fix from PR #615 + accept the macos-latest red bar via a documented "known regression" entry in the smoke harness, NOT silent.

### B. Vendor a patched fastembed inside moflo (interim while A is pending)

Fork fastembed locally (e.g., `vendor/fastembed/`) with the pin relaxed, depend on the local copy via `file:`. When upstream A merges, switch back to the npm registry version.

- **Pros:** unblocks macOS without waiting on upstream.
- **Cons:** maintenance burden, security update tracking falls on us.

### C. Replace fastembed with a thin moflo-owned wrapper around onnxruntime-node

Drop fastembed entirely, write a small wrapper in `src/modules/cli/src/embeddings/` that uses onnxruntime-node + a tokenizer of our choice directly.

- **Pros:** eliminates the transitive-pin coupling permanently; we control the ORT version end-to-end.
- **Cons:** larger lift; have to re-implement the all-MiniLM-L6-v2 download/cache/inference path; risk regressions in the embedding pipeline that's currently working.

### D. Refined Attempt-2 surgery (LAST resort — only if A/B/C all blocked)

If we MUST do postinstall surgery, the three problems from Attempt 2 each need an explicit answer:

1. **Install-size:** either bump the smoke threshold AND ship a follow-up that aggressively trims both ORT copies' multi-platform binaries before the size check runs, OR teach `findOrtPackages` that "moflo declares ORT" should NOT count as a shared declarer.
2. **Windows EPERM:** either use a Node-side approach that doesn't require directory removal (e.g., rename `package.json` so Node resolution skips the dir, then aggressively `rm` just `bin/` and `lib/` subdirs which are usually unlocked), OR accept Windows local dev as a known-friction point.
3. **Windows fastembed install.js crash:** investigate whether `onnxruntime-node@1.21.0`'s install.js fails on Linux/macOS too. If yes, this attempt is dead. If only Windows, see #2.

**Whoever picks D up should write the analysis in the ticket BEFORE coding, not during.**

## What's safe to keep from PR #615

The napi-v3 → napi-v6 dynamic-discovery refactor of `scripts/prune-native-binaries.mjs` and `harness/consumer-smoke/lib/checks.mjs` is a real fix for a future regression class (the prune was about to silently no-op the moment any future fastembed bump shipped). 19 unit tests, 0 regressions on Linux/macOS/Windows.

The `onnxruntime-node@>=1.24.3` entry in the `overrides` block is dev-install-only theatre for this issue but is harmless and may become useful once approach A lands.

**Recommended action on PR #615:** rescope its title and body to "fix(scripts): handle ORT 1.24+ napi-v6 binary layout in pruner" (no longer claims to fix the macOS crash), keep #613 open, link the new scope to a separate issue or scrap the #613 closing reference.

## Acceptance Criteria (for whoever picks this up)

- [ ] In the ticket, explicitly answer: which path (A / B / C / D) is being taken and why.
- [ ] Before writing code, document: how does this fix reach the consumer install (NOT just the dev install)? What's the verification step?
- [ ] Before writing code, document: cross-platform behavior on Linux + macOS + Windows. Specifically address Windows file locks and macOS-only dynamic linker behavior.
- [ ] `Consumer install smoke / macos-latest` returns to green on the PR.
- [ ] `Consumer install smoke / ubuntu-latest` and `/ windows-latest` remain green.
- [ ] No new entries in the path-resolution stderr scan.
- [ ] `npm test` passes on Linux.
- [ ] The `tokenizer-subpackage` and `ort-pruned` smoke checks remain green.

## Severity

HIGH — `macos-latest` smoke is permanently red, which violates `feedback_broken_window_theory.md` and means our consumer-install guarantees on macOS are unverified for every PR. PRs that make zero macOS-relevant changes (e.g. PR #612, a pure-deletion + docs diff) are still being blocked / require admin-merge — broken-window noise.

## References

- Upstream issue: https://github.com/microsoft/onnxruntime/issues/24579
- Upstream fix PR: https://github.com/microsoft/onnxruntime/pull/26445 (merged 2025-10-31)
- onnxruntime-node v1.24.1 release: https://github.com/microsoft/onnxruntime/releases/tag/v1.24.1
- fastembed (the package with the exact pin): https://github.com/Anush008/fastembed-js
- PR #615 (Attempt 1, partial fix only): https://github.com/eric-cielo/moflo/pull/615
- Failed run on main: https://github.com/eric-cielo/moflo/actions/runs/24930845358
- Failed run on PR #615: https://github.com/eric-cielo/moflo/actions/runs/24932238217
- `feedback_broken_window_theory.md`
- `feedback_cross_platform_mandatory.md` — macOS is a first-class supported platform
- `feedback_call_out_fix_or_file.md`

## Out-of-scope

- This is not caused by, and cannot be fixed by, any of the workspace-collapse stories (#590-#602) — those are pure import-rewrite / deletion PRs.
- Process-gate work (forcing cross-platform + consumer analysis before native-dep edits) was scoped during this investigation but is OUT OF SCOPE for this ticket and tracked separately.
- Filing this from PR #612 (deletion of `@moflo/plugins`) per the "fix or file" rule (`feedback_call_out_fix_or_file.md`) — admin-merging that PR despite this red signal because the cause is unrelated.

--- a/harness/consumer-smoke/lib/checks.mjs
+++ b/harness/consumer-smoke/lib/checks.mjs
@@ -5,12 +5,12 @@
  */
 
 import { existsSync, mkdirSync, writeFileSync, readdirSync, rmSync, statSync, readFileSync } from 'node:fs';
-import { join, relative } from 'node:path';
+import { basename, join, relative } from 'node:path';
 
 import { run, runNode, flo, NPM_CMD, getStderrSamples } from './proc.mjs';
 import { section, record, recordExit, log } from './report.mjs';
 import { findOrphans } from '../../../scripts/clean-dist.mjs';
-import { findOrtPackages } from '../../../scripts/prune-native-binaries.mjs';
+import { findOrtPackages, listNapiDirs } from '../../../scripts/prune-native-binaries.mjs';
 
 // Epic #501 acceptance criterion: a fresh `npm install moflo` consumer sees
 // zero of the following packages anywhere in its dep tree.
@@ -641,31 +641,43 @@ function logTopOffenders(nodeModulesDir) {
   }
 }
 
-/** Inspect one onnxruntime-node install; return a problem string or null. */
+/**
+ * Inspect one onnxruntime-node install; return a problem string or null.
+ *
+ * ORT names this directory after the N-API ABI it was built against — `napi-v3`
+ * in 1.21, `napi-v6` in 1.24+ — so we discover whichever subdir(s) the install
+ * actually ships (via the same helper the pruner uses) and assert each one is
+ * pruned to the current platform/arch.
+ */
 function inspectOrtInstall(ortDir, consumerDir) {
-  const napi = join(ortDir, 'bin', 'napi-v3');
-  if (!existsSync(napi)) return null; // future layout change, not a prune failure
+  const napiDirs = listNapiDirs(join(ortDir, 'bin'));
+  if (napiDirs.length === 0) return null; // future layout change, not a prune failure
+
   const rel = relative(consumerDir, ortDir);
   const dirNames = (d) => readdirSync(d, { withFileTypes: true })
     .filter(e => e.isDirectory())
     .map(e => e.name);
 
-  const platforms = dirNames(napi);
-  const platExtras = platforms.filter(p => p !== process.platform);
-  if (platExtras.length > 0) return `${rel}: extra platforms present — ${platExtras.join(', ')}`;
-  if (!platforms.includes(process.platform)) return `${rel}: current platform dir missing (${process.platform})`;
+  for (const napi of napiDirs) {
+    const napiName = basename(napi);
+    const platforms = dirNames(napi);
+    const platExtras = platforms.filter(p => p !== process.platform);
+    if (platExtras.length > 0) return `${rel}: extra platforms in ${napiName} — ${platExtras.join(', ')}`;
+    if (!platforms.includes(process.platform)) return `${rel}: current platform dir missing in ${napiName} (${process.platform})`;
 
-  const archs = dirNames(join(napi, process.platform));
-  const archExtras = archs.filter(a => a !== process.arch);
-  if (archExtras.length > 0) return `${rel}: extra archs under ${process.platform} — ${archExtras.join(', ')}`;
-  if (!archs.includes(process.arch)) return `${rel}: current arch dir missing (${process.arch})`;
+    const archs = dirNames(join(napi, process.platform));
+    const archExtras = archs.filter(a => a !== process.arch);
+    if (archExtras.length > 0) return `${rel}: extra archs under ${napiName}/${process.platform} — ${archExtras.join(', ')}`;
+    if (!archs.includes(process.arch)) return `${rel}: current arch dir missing in ${napiName}/${process.platform} (${process.arch})`;
+  }
   return null;
 }
 
 /**
  * After postinstall, every onnxruntime-node install should retain only the
- * current `<platform>/<arch>` subtree under `bin/napi-v3/`. Hard-fail on any
- * non-current platform directory or any extra arch under the kept platform.
+ * current `<platform>/<arch>` subtree under each `bin/napi-v<N>/` directory.
+ * Hard-fail on any non-current platform directory or any extra arch under the
+ * kept platform.
  */
 export function verifyPrunedBinaries(consumerDir) {
   section('Verify onnxruntime-node binaries pruned to current platform');

--- a/package.json
+++ b/package.json
@@ -92,6 +92,7 @@
   },
   "overrides": {
     "hono": ">=4.11.4",
+    "onnxruntime-node": ">=1.24.3",
     "picomatch": ">=2.3.2",
     "protobufjs": ">=7.5.5",
     "tar": ">=7.5.11"

--- a/scripts/prune-native-binaries.mjs
+++ b/scripts/prune-native-binaries.mjs
@@ -3,10 +3,12 @@
  * Postinstall native-binary pruner for consumer moflo installs.
  *
  * Trims `onnxruntime-node`'s multi-platform binary bundle under
- * `bin/napi-v3/<platform>/<arch>/` down to just the current combination,
+ * `bin/napi-v<N>/<platform>/<arch>/` down to just the current combination,
  * and also strips GPU-only provider libraries (CUDA ~327 MB, TensorRT,
  * DirectML ~18 MB) that fastembed never loads. Reclaims ~340 MB per Linux
- * install and ~150 MB per Windows install.
+ * install and ~150 MB per Windows install. The `napi-v<N>` directory name
+ * tracks the upstream N-API ABI (v3 in ORT 1.21, v6 in ORT 1.24+); the
+ * pruner discovers it dynamically so future bumps don't silently no-op.
  *
  * Ownership-scoped: only prunes `onnxruntime-node` copies that fastembed
  * owns via its dependency graph. Foreign sibling ORT installs (e.g. an
@@ -38,17 +40,24 @@ const FASTEMBED_OWNER = 'fastembed';
 const ORT_DEP_FIELDS = ['dependencies', 'optionalDependencies', 'peerDependencies'];
 // GPU-only provider libraries that ship in onnxruntime-node's binary bundle but
 // are never loaded by fastembed (CPU inference). Matches files like:
-//   libonnxruntime_providers_cuda.so      (linux — 327 MB for ORT 1.21)
+//   libonnxruntime_providers_cuda.so      (linux — 327 MB for ORT 1.21+)
 //   libonnxruntime_providers_tensorrt.so  (linux)
 //   libonnxruntime_providers_shared.so    (linux — GPU-loader helper only)
 //   onnxruntime_providers_*.dll           (windows variant, same role)
 //   DirectML.dll                          (windows DirectX GPU — 18 MB)
 // The CPU runtime (`libonnxruntime.so.1*`, `onnxruntime.dll`, `libonnxruntime.*.dylib`)
 // and the Node binding (`onnxruntime_binding.node`) never match and are preserved.
-// Other GPU providers (openvino, rocm, qnn) aren't shipped in ORT 1.21; revisit
-// on version bumps if they appear.
+// Other GPU providers (openvino, rocm, qnn) aren't shipped in ORT 1.21–1.24;
+// revisit on version bumps if they appear.
 const GPU_PROVIDER_FILE_PATTERN =
   /^(lib)?onnxruntime_providers_(cuda|tensorrt|shared)(\.|$)|^directml\.dll$/i;
+
+// Matches ORT's per-N-API-version subdirectory under `bin/`. ORT 1.21 ships
+// `napi-v3`, ORT 1.24 ships `napi-v6`, and we expect this number to keep
+// climbing — discovering it dynamically avoids silent no-ops on future bumps.
+// Exported so the consumer-smoke harness can assert on the same shape without
+// re-defining the literal.
+export const NAPI_DIR_PATTERN = /^napi-v\d+$/;
 
 function info(msg) { console.log(`moflo:prune ${msg}`); }
 function warn(msg) { console.warn(`moflo:prune warn: ${msg}`); }
@@ -283,13 +292,13 @@ export function removeDir(dir, { rm = rmSync, measure = false } = {}) {
 
 /**
  * Plant a zero-byte `libonnxruntime_providers_cuda.so` stub in the linux/x64
- * arch dir. `onnxruntime-node@1.21`'s own postinstall downloads a 327 MB CUDA
+ * arch dir. `onnxruntime-node`'s own postinstall downloads a 327 MB CUDA
  * provider tarball *after* moflo's postinstall runs (npm orders file:-tarball
- * roots before their deps). The download is gated on `!CUDA_DLL_EXISTS`, so
- * pre-creating the path as an empty file short-circuits the fetch entirely.
- * No-op when the user opts in to CUDA via `ONNXRUNTIME_NODE_INSTALL_CUDA=true`
- * or when the stub already exists. See `node_modules/onnxruntime-node/script/install.js`
- * for the upstream check.
+ * roots before their deps). The download is gated on `existsSync(filepath)`
+ * for each manifested file, so pre-creating the path as an empty file
+ * short-circuits the fetch entirely. No-op when the user opts in to CUDA via
+ * `ONNXRUNTIME_NODE_INSTALL_CUDA=true` or when the stub already exists. See
+ * `node_modules/onnxruntime-node/script/install.js` for the upstream check.
  */
 export function plantCudaStub(archDir, { keepPlatform, keepArch, env = process.env }) {
   if (keepPlatform !== 'linux' || keepArch !== 'x64') return false;
@@ -334,12 +343,28 @@ export function pruneGpuProviders(archDir, { rm = rmSync, measure = false } = {}
 }
 
 /**
- * Prune `bin/napi-v3/<platform>/<arch>/` from one onnxruntime-node package,
+ * Discover every `bin/napi-v<N>` subdir present in an ORT install. ORT 1.21
+ * ships only `napi-v3`, ORT 1.24 ships only `napi-v6`, but a transitional
+ * release could ship both — handle each independently rather than guessing.
+ * Returns absolute paths; missing/unreadable `binDir` yields an empty array.
+ */
+export function listNapiDirs(binDir) {
+  let entries;
+  try { entries = readdirSync(binDir, { withFileTypes: true }); }
+  catch { return []; }
+  return entries
+    .filter((e) => e.isDirectory() && NAPI_DIR_PATTERN.test(e.name))
+    .map((e) => join(binDir, e.name));
+}
+
+/**
+ * Prune `bin/napi-v<N>/<platform>/<arch>/` from one onnxruntime-node package,
  * keeping only the current combination. Also strips GPU-only provider
  * libraries (`libonnxruntime_providers_{cuda,tensorrt,shared}.*`, `DirectML.dll`)
  * from the kept arch dir — fastembed runs on the CPU provider only, so these
- * are dead weight on every host (ORT 1.21 ships a 327 MB CUDA provider on
- * linux/x64 that nothing in our graph loads).
+ * are dead weight on every host (ORT 1.21+ ships a 327 MB CUDA provider on
+ * linux/x64 that nothing in our graph loads). Iterates every `napi-v<N>`
+ * directory present so the pruner survives N-API ABI bumps.
  */
 export function pruneOrtPackage(ortDir, {
   keepPlatform = process.platform,
@@ -348,33 +373,37 @@ export function pruneOrtPackage(ortDir, {
   measure = false,
   env = process.env,
 } = {}) {
-  const napiDir = join(ortDir, 'bin', 'napi-v3');
-  let platforms;
-  try { platforms = readdirSync(napiDir, { withFileTypes: true }); }
-  catch { return 0; }
+  const napiDirs = listNapiDirs(join(ortDir, 'bin'));
+  if (napiDirs.length === 0) return 0;
 
   let bytes = 0;
-  for (const p of platforms) {
-    if (!p.isDirectory()) continue;
-    const pDir = join(napiDir, p.name);
-    if (p.name !== keepPlatform) {
-      bytes += removeDir(pDir, { rm, measure });
-      continue;
-    }
-    let archs;
-    try { archs = readdirSync(pDir, { withFileTypes: true }); }
-    catch (err) { warn(`cannot read ${pDir}: ${err.message}`); continue; }
-    for (const a of archs) {
-      if (!a.isDirectory()) continue;
-      if (a.name === keepArch) {
-        const archDir = join(pDir, a.name);
-        bytes += pruneGpuProviders(archDir, { rm, measure });
-        // Plant stub AFTER prune so the upstream ORT postinstall (which runs
-        // later on linux/x64) sees `CUDA_DLL_EXISTS` and skips the fetch.
-        plantCudaStub(archDir, { keepPlatform, keepArch, env });
+  for (const napiDir of napiDirs) {
+    let platforms;
+    try { platforms = readdirSync(napiDir, { withFileTypes: true }); }
+    catch { continue; }
+
+    for (const p of platforms) {
+      if (!p.isDirectory()) continue;
+      const pDir = join(napiDir, p.name);
+      if (p.name !== keepPlatform) {
+        bytes += removeDir(pDir, { rm, measure });
         continue;
       }
-      bytes += removeDir(join(pDir, a.name), { rm, measure });
+      let archs;
+      try { archs = readdirSync(pDir, { withFileTypes: true }); }
+      catch (err) { warn(`cannot read ${pDir}: ${err.message}`); continue; }
+      for (const a of archs) {
+        if (!a.isDirectory()) continue;
+        if (a.name === keepArch) {
+          const archDir = join(pDir, a.name);
+          bytes += pruneGpuProviders(archDir, { rm, measure });
+          // Plant stub AFTER prune so the upstream ORT postinstall (which
+          // runs later on linux/x64) sees the CUDA file and skips the fetch.
+          plantCudaStub(archDir, { keepPlatform, keepArch, env });
+          continue;
+        }
+        bytes += removeDir(join(pDir, a.name), { rm, measure });
+      }
     }
   }
   return bytes;

--- a/tests/scripts/prune-native-binaries.test.mts
+++ b/tests/scripts/prune-native-binaries.test.mts
@@ -42,22 +42,38 @@ afterEach(() => {
   if (existsSync(tmp)) rmSync(tmp, { recursive: true, force: true, maxRetries: 3, retryDelay: 50 });
 });
 
-/** Build an `onnxruntime-node/bin/napi-v3/<plat>/<arch>/` fixture tree. */
+/**
+ * Build an `onnxruntime-node/bin/napi-v<N>/<plat>/<arch>/` fixture tree.
+ * `napiVersion` defaults to `3` (matching ORT 1.21's layout); pass `6` to
+ * model ORT 1.24+, or call `addNapiTier` to plant additional N-API dirs in
+ * the same package.
+ */
 function buildOrtTree(
   root: string,
   combos: Array<{ platform: string; arch: string; bytes?: number }>,
+  { napiVersion = 3 }: { napiVersion?: number } = {},
 ): string {
   const ortDir = join(root, 'onnxruntime-node');
-  const napi = join(ortDir, 'bin', 'napi-v3');
+  addNapiTier(ortDir, combos, napiVersion);
+  // Non-binary sibling file to confirm we don't touch anything outside bin/.
+  writeFileSync(join(ortDir, 'package.json'), '{"name":"onnxruntime-node"}');
+  return ortDir;
+}
+
+/** Plant or extend a `bin/napi-v<N>/` tier inside an existing ORT fixture. */
+function addNapiTier(
+  ortDir: string,
+  combos: Array<{ platform: string; arch: string; bytes?: number }>,
+  napiVersion: number,
+): string {
+  const napi = join(ortDir, 'bin', `napi-v${napiVersion}`);
   mkdirSync(napi, { recursive: true });
   for (const { platform, arch, bytes = 1024 } of combos) {
     const leaf = join(napi, platform, arch);
     mkdirSync(leaf, { recursive: true });
     writeFileSync(join(leaf, 'onnxruntime_binding.node'), Buffer.alloc(bytes));
   }
-  // Non-binary sibling file to confirm we don't touch anything outside napi-v3
-  writeFileSync(join(ortDir, 'package.json'), '{"name":"onnxruntime-node"}');
-  return ortDir;
+  return napi;
 }
 
 describe('findOrtPackages (fastembed-owned only)', () => {
@@ -305,19 +321,58 @@ describe('pruneOrtPackage', () => {
     expect(existsSync(join(ortDir, 'bin', 'napi-v3', 'linux', 'x64'))).toBe(true);
   });
 
-  it('returns 0 when bin/napi-v3 is missing (future layout change)', () => {
+  it('returns 0 when no bin/napi-v<N>/ subdir exists (future layout change)', () => {
     const ortDir = join(tmp, 'onnxruntime-node');
     mkdirSync(ortDir, { recursive: true });
     expect(pruneOrtPackage(ortDir)).toBe(0);
   });
 
-  it('preserves sibling files outside bin/napi-v3', () => {
+  it('returns 0 when bin/ has only non-napi siblings', () => {
+    const ortDir = join(tmp, 'onnxruntime-node');
+    mkdirSync(join(ortDir, 'bin', 'something-else'), { recursive: true });
+    expect(pruneOrtPackage(ortDir)).toBe(0);
+  });
+
+  it('preserves sibling files outside bin/napi-v<N>/', () => {
     const ortDir = buildOrtTree(tmp, [
       { platform: 'linux', arch: 'x64' },
       { platform: 'win32', arch: 'x64' },
     ]);
     pruneOrtPackage(ortDir, { keepPlatform: 'linux', keepArch: 'x64' });
     expect(existsSync(join(ortDir, 'package.json'))).toBe(true);
+  });
+
+  it('prunes a napi-v6 layout (ORT 1.24+) the same way as napi-v3', () => {
+    // Regression test for issue #613: ORT 1.24's `bin/napi-v6/` layout was
+    // silently no-op'd when the pruner had `napi-v3` hardcoded. Verifies the
+    // dynamic discovery handles whichever ABI dir the install actually ships.
+    const ortDir = buildOrtTree(
+      tmp,
+      ALL_PLATFORMS.flatMap((p) => ALL_ARCHES.map((a) => ({ platform: p, arch: a }))),
+      { napiVersion: 6 },
+    );
+
+    pruneOrtPackage(ortDir, { keepPlatform: 'darwin', keepArch: 'arm64' });
+
+    const napi = join(ortDir, 'bin', 'napi-v6');
+    expect(readdirSync(napi)).toEqual(['darwin']);
+    expect(readdirSync(join(napi, 'darwin'))).toEqual(['arm64']);
+  });
+
+  it('prunes both napi-v3 and napi-v6 when an install ships both', () => {
+    // Defensive: a transitional ORT release could ship both ABI dirs at once.
+    // The pruner walks each independently rather than picking one.
+    const combos = ALL_PLATFORMS.flatMap((p) => ALL_ARCHES.map((a) => ({ platform: p, arch: a })));
+    const ortDir = buildOrtTree(tmp, combos, { napiVersion: 3 });
+    addNapiTier(ortDir, combos, 6);
+
+    pruneOrtPackage(ortDir, { keepPlatform: 'linux', keepArch: 'x64' });
+
+    for (const napiName of ['napi-v3', 'napi-v6']) {
+      const napi = join(ortDir, 'bin', napiName);
+      expect(readdirSync(napi)).toEqual(['linux']);
+      expect(readdirSync(join(napi, 'linux'))).toEqual(['x64']);
+    }
   });
 
   it('retries once on EBUSY then swallows persistent failures', () => {

